wp-admin/js/widgets/text-widgets.js

```diff
@@ -0,0 +1,259 @@
+	'use strict';
+
+	var component = {};
+
+	/**
+	 * Text widget control.
+	 *
+	 * @class TextWidgetControl
+	 * @constructor
+	 * @abstract
+	 */
+	component.TextWidgetControl = Backbone.View.extend({
+
+		/**
+		 * View events.
+		 *
+		 * @type {Object}
+		 */
+		events: {},
+
+		/**
+		 * Initialize.
+		 *
+		 * @param {Object}         options - Options.
+		 * @param {Backbone.Model} options.model - Model.
+		 * @param {jQuery}         options.el - Control container element.
+		 * @returns {void}
+		 */
+		initialize: function initialize( options ) {
+			var control = this;
+
+			Backbone.View.prototype.initialize.call( control, options );
+
+			if ( ! control.el ) {
+				throw new Error( 'Missing options.el' );
+			}
+		},
+
+		/**
+		 * Render template.
+		 *
+		 * @returns {void}
+		 */
+		render: function render() {
+			var control = this, changeDebounceDelay = 1000, iframeKeepAliveInterval = 1000, id, textarea;
+			textarea = control.$el.find( 'textarea:first' );
+			id = textarea.attr( 'id' );
+
+			/**
+			 * Build (or re-build) an the visual editor.
+			 *
+			 * @returns {void}
+			 */
+			function buildEditor() {
+				var editor, wrap, onChanged, dirty, triggerChangeIfDirty;
+
+				// Destroy any existing editor so that it can be re-initialized after a widget-updated event.
+				if ( tinymce.get( id ) ) {
+					delete tinymce.editors[ id ];
+					tinymce.remove( '#' + id );
+				}
+
+				// Remove any previous QuickTags created in the previous render.
+				if ( QTags.instances[ id ] ) {
+					delete QTags.instances[ id ];
+				}
+
+				// Unwrap the textarea to its original location in the DOM.
+				wrap = $( '#wp-' + id + '-wrap' );
+				if ( wrap.length ) {
+					textarea.show();
+					wrap.replaceWith( textarea );
+				}
+
+				wp.editor.initialize( id, {
+					tinymce: {
+						wpautop: true
+					},
+					quicktags: true
+				} );
+				QTags._buttonsInit(); // @todo Remove once <https://core.trac.wordpress.org/ticket/35760#comment:28> is resolved.
+
+				editor = window.tinymce.get( id );
+				if ( editor.initialized ) {
+					watchForDestroyedBody( control.$el.find( 'iframe' )[0] );
+				} else {
+					editor.on( 'init', function() {
+						watchForDestroyedBody( control.$el.find( 'iframe' )[0] );
+					} );
+				}
+
+				dirty = false;
+				onChanged = function() {
+					dirty = true;
+				};
+				triggerChangeIfDirty = function() {
+					if ( dirty ) {
+						editor.save();
+						textarea.trigger( 'change' );
+						dirty = false;
+					}
+				};
+				editor.on( 'focus', function() {
+					editor.on( 'change', onChanged );
+				} );
+				editor.on( 'change', _.debounce( triggerChangeIfDirty, changeDebounceDelay ) );
+				editor.on( 'blur', function() {
+					editor.off( 'change', onChanged );
+					triggerChangeIfDirty();
+				} );
+			}
+
+			/**
+			 * Watch an iframe for the destruction of its TinyMCE contenteditable contents.
+			 *
+			 * @todo There may be a better way to listen for an iframe being destroyed.
```

Comment:
Any idea for a better way to listen for the destruction of an iframe's content document?